File: src/lib/seed.ts
import type {
  Album,
  AppNotification,
  ArtistPayout,
  Playlist,
  SubscriptionPrices,
  Ticket,
  Track,
  User,
} from "./types";
import { todayISO } from "./utils";

const cover = (seed: string) =>
  `https://picsum.photos/seed/${seed}/400/400`;

/** Royalty-free sample tracks for demo playback */
const AUDIO = {
  a: "https://www.soundhelix.com/examples/mp3/SoundHelix-Song-1.mp3",
  b: "https://www.soundhelix.com/examples/mp3/SoundHelix-Song-2.mp3",
  c: "https://www.soundhelix.com/examples/mp3/SoundHelix-Song-3.mp3",
  d: "https://www.soundhelix.com/examples/mp3/SoundHelix-Song-4.mp3",
  e: "https://www.soundhelix.com/examples/mp3/SoundHelix-Song-5.mp3",
};

const defaultSettings = {
  notificationsEnabled: true,
  volume: 0.8,
  language: "fa" as const,
};

export const DEMO_PASSWORDS = "demo1234";

export function createSeedUsers(): User[] {
  const day = todayISO();
  return [
    {
      id: "user_listener_basic",
      role: "listener",
      displayName: "سارا شنونده",
      username: "sara_basic",
      email: "sara@demo.com",
      password: DEMO_PASSWORDS,
      subscriptionTier: "basic",
      followers: ["user_listener_gold"],
      following: ["user_artist_verified"],
      dailyStreamCount: 12,
      dailyStreamDate: day,
      settings: { ...defaultSettings },
      gender: "female",
      birthDate: "1998-05-12",
      createdAt: "2026-01-01T10:00:00.000Z",
    },
    {
      id: "user_listener_silver",
      role: "listener",
      displayName: "علی نقره‌ای",
      username: "ali_silver",
      email: "ali@demo.com",
      password: DEMO_PASSWORDS,
      avatar: cover("ali"),
      subscriptionTier: "silver",
      subscriptionExpiresAt: "2026-12-01",
      followers: [],
      following: ["user_artist_verified"],
      dailyStreamCount: 40,
      dailyStreamDate: day,
      settings: { ...defaultSettings },
      gender: "male",
      birthDate: "1995-03-20",
      createdAt: "2026-01-02T10:00:00.000Z",
    },
    {
      id: "user_listener_gold",
      role: "listener",
      displayName: "نیما طلایی",
      username: "nima_gold",
      email: "nima@demo.com",
      password: DEMO_PASSWORDS,
      avatar: cover("nima"),
      subscriptionTier: "gold",
      subscriptionExpiresAt: "2027-01-01",
      followers: ["user_listener_basic"],
      following: ["user_artist_verified", "user_artist_pending"],
      dailyStreamCount: 100,
      dailyStreamDate: day,
      settings: { ...defaultSettings },
      gender: "male",
      birthDate: "1992-11-08",
      createdAt: "2026-01-03T10:00:00.000Z",
    },
    {
      id: "user_artist_verified",
      role: "artist",
      displayName: "آوا مهر",
      username: "ava_mehr",
      email: "ava@demo.com",
      password: DEMO_PASSWORDS,
      avatar: cover("ava"),
      subscriptionTier: "gold",
      followers: ["user_listener_basic", "user_listener_silver", "user_listener_gold"],
      following: [],
      dailyStreamCount: 5,
      dailyStreamDate: day,
      settings: { ...defaultSettings },
      artistProfile: {
        bio: "خواننده و آهنگساز مستقل با تمرکز روی پاپ و الکترونیک.",
        verified: true,
        status: "approved",
        sampleWorks: "نمونه‌کارهای منتشرشده روی پلتفرم",
        stageName: "آوا مهر",
      },
      createdAt: "2025-12-01T10:00:00.000Z",
    },
    {
      id: "user_artist_pending",
      role: "artist",
      displayName: "کاوه نوپا",
      username: "kaveh_new",
      email: "kaveh@demo.com",
      password: DEMO_PASSWORDS,
      subscriptionTier: "basic",
      followers: [],
      following: [],
      dailyStreamCount: 0,
      dailyStreamDate: day,
      settings: { ...defaultSettings },
      artistProfile: {
        bio: "هنرمند تازه‌کار در انتظار تأیید.",
        verified: false,
        status: "pending",
        sampleWorks: "لینک دمو / فایل‌های نمونه",
        stageName: "کاوه نوپا",
      },
      createdAt: "2026-07-01T10:00:00.000Z",
    },
    {
      id: "user_support",
      role: "support",
      displayName: "پشتیبان ملودی",
      username: "support1",
      email: "support@demo.com",
      password: DEMO_PASSWORDS,
      subscriptionTier: "gold",
      followers: [],
      following: [],
      dailyStreamCount: 0,
      dailyStreamDate: day,
      settings: { ...defaultSettings },
      createdAt: "2025-11-01T10:00:00.000Z",
    },
    {
      id: "user_admin",
      role: "admin",
      displayName: "مدیر سامانه",
      username: "admin",
      email: "admin@demo.com",
      password: DEMO_PASSWORDS,
      avatar: cover("admin"),
      subscriptionTier: "gold",
      followers: [],
      following: [],
      dailyStreamCount: 0,
      dailyStreamDate: day,
      settings: { ...defaultSettings },
      createdAt: "2025-10-01T10:00:00.000Z",
    },
  ];
}

export function createSeedTracks(): Track[] {
  return [
    {
      id: "track_1",
      title: "سپیده‌دم",
      artistIds: ["user_artist_verified"],
      albumId: "album_1",
      cover: cover("track1"),
      audioUrl: AUDIO.a,
      lyrics: "در سپیده‌دم\nصدای شهر آرام می‌گیرد\nو دل به راه می‌رود",
      genre: "پاپ",
      year: 2025,
      streams: 15200,
      listeners: 4200,
      duration: 372,
      earlyAccess: false,
      isSingle: false,
      createdAt: "2025-06-01T10:00:00.000Z",
    },
    {
      id: "track_2",
      title: "شب‌های تهران",
      artistIds: ["user_artist_verified"],
      albumId: "album_1",
      cover: cover("track2"),
      audioUrl: AUDIO.b,
      lyrics: "شب‌های تهران\nنور نئون و صدای باران",
      genre: "پاپ",
      year: 2025,
      streams: 22100,
      listeners: 6100,
      duration: 401,
      earlyAccess: false,
      isSingle: false,
      createdAt: "2025-06-01T10:00:00.000Z",
    },
    {
      id: "track_3",
      title: "موج آرام",
      artistIds: ["user_artist_verified"],
      albumId: "album_1",
      cover: cover("track3"),
      audioUrl: AUDIO.c,
      genre: "الکترونیک",
      year: 2025,
      streams: 9800,
      listeners: 3000,
      duration: 350,
      earlyAccess: false,
      isSingle: false,
      createdAt: "2025-06-02T10:00:00.000Z",
    },
    {
      id: "track_4",
      title: "تک‌آهنگ طلایی",
      artistIds: ["user_artist_verified"],
      cover: cover("track4"),
      audioUrl: AUDIO.d,
      lyrics: "این تک‌آهنگ زودهنگام است",
      genre: "ایندی",
      year: 2026,
      streams: 1200,
      listeners: 800,
      duration: 310,
      earlyAccess: true,
      isSingle: true,
      createdAt: "2026-07-10T10:00:00.000Z",
    },
    {
      id: "track_5",
      title: "پرواز",
      artistIds: ["user_artist_verified"],
      albumId: "album_2",
      cover: cover("track5"),
      audioUrl: AUDIO.e,
      genre: "راک",
      year: 2024,
      streams: 45000,
      listeners: 12000,
      duration: 290,
      earlyAccess: false,
      isSingle: false,
      createdAt: "2024-09-01T10:00:00.000Z",
    },
  ];
}

export function createSeedAlbums(): Album[] {
  return [
    {
      id: "album_1",
      title: "شهر خاموش",
      artistIds: ["user_artist_verified"],
      cover: cover("album1"),
      year: 2025,
      trackIds: ["track_1", "track_2", "track_3"],
      genre: "پاپ",
      earlyAccess: false,
      createdAt: "2025-06-01T10:00:00.000Z",
    },
    {
      id: "album_2",
      title: "ارتفاع",
      artistIds: ["user_artist_verified"],
      cover: cover("album2"),
      year: 2024,
      trackIds: ["track_5"],
      genre: "راک",
      earlyAccess: false,
      createdAt: "2024-09-01T10:00:00.000Z",
    },
  ];
}

export function createSeedPlaylists(): Playlist[] {
  return [
    {
      id: "pl_1",
      ownerId: "user_listener_basic",
      name: "تمرکز صبح",
      trackIds: ["track_1", "track_3"],
      cover: cover("pl1"),
      createdAt: "2026-02-01T10:00:00.000Z",
      updatedAt: "2026-02-01T10:00:00.000Z",
    },
    {
      id: "pl_2",
      ownerId: "user_listener_gold",
      name: "شب‌گردی",
      trackIds: ["track_2", "track_4", "track_5"],
      cover: cover("pl2"),
      createdAt: "2026-03-01T10:00:00.000Z",
      updatedAt: "2026-03-01T10:00:00.000Z",
    },
  ];
}

export function createSeedNotifications(): AppNotification[] {
  return [
    {
      id: "notif_1",
      userId: "user_listener_basic",
      title: "مهلت اشتراک",
      body: "اشتراک پایه شما رایگان است؛ برای امکانات بیشتر ارتقا دهید.",
      read: false,
      href: "/settings",
      createdAt: "2026-07-20T10:00:00.000Z",
    },
    {
      id: "notif_2",
      userId: "user_listener_gold",
      title: "اثر جدید",
      body: "آوا مهر تک‌آهنگ جدیدی منتشر کرد.",
      read: false,
      href: "/browse",
      createdAt: "2026-07-21T10:00:00.000Z",
    },
    {
      id: "notif_3",
      userId: "user_artist_pending",
      title: "در انتظار تأیید",
      body: "درخواست حساب هنرمندی شما در حال بررسی است.",
      read: true,
      createdAt: "2026-07-01T12:00:00.000Z",
    },
    {
      id: "notif_4",
      userId: "user_support",
      title: "تیکت جدید",
      body: "یک تیکت پشتیبانی جدید ثبت شد.",
      read: false,
      href: "/admin/tickets",
      createdAt: "2026-07-22T08:00:00.000Z",
    },
    {
      id: "notif_5",
      userId: "user_admin",
      title: "درخواست هنرمند",
      body: "کاوه نوپا درخواست تأیید حساب داده است.",
      read: false,
      href: "/admin/artist-requests",
      createdAt: "2026-07-22T09:00:00.000Z",
    },
  ];
}

export function createSeedTickets(): Ticket[] {
  return [
    {
      id: "ticket_1",
      userId: "user_listener_basic",
      subject: "مشکل پخش آهنگ",
      status: "open",
      messages: [
        {
          id: "tm_1",
          senderId: "user_listener_basic",
          body: "بعضی آهنگ‌ها بعد از چند ثانیه قطع می‌شوند.",
          createdAt: "2026-07-22T07:00:00.000Z",
        },
      ],
      createdAt: "2026-07-22T07:00:00.000Z",
    },
    {
      id: "ticket_2",
      userId: "user_listener_silver",
      subject: "سوال درباره دانلود",
      status: "answered",
      messages: [
        {
          id: "tm_2",
          senderId: "user_listener_silver",
          body: "دانلود آهنگ از کجا فعال می‌شود؟",
          createdAt: "2026-07-18T10:00:00.000Z",
        },
        {
          id: "tm_3",
          senderId: "user_support",
          body: "از منوی کارت آهنگ گزینه دانلود را بزنید (اشتراک نقره‌ای و بالاتر).",
          createdAt: "2026-07-18T12:00:00.000Z",
        },
      ],
      createdAt: "2026-07-18T10:00:00.000Z",
    },
  ];
}

export function createSeedPayouts(): ArtistPayout[] {
  return [
    {
      id: "pay_1",
      artistId: "user_artist_verified",
      uniqueListeners: 8200,
      streams: 94000,
      amount: 4500000,
      paymentStatus: "pending",
      month: "2026-07",
    },
  ];
}

export function createSeedPrices(): SubscriptionPrices {
  return { silver: 99000, gold: 199000 };
}

export interface SeedData {
  users: User[];
  tracks: Track[];
  albums: Album[];
  playlists: Playlist[];
  notifications: AppNotification[];
  tickets: Ticket[];
  payouts: ArtistPayout[];
  prices: SubscriptionPrices;
}

export function createFullSeed(): SeedData {
  return {
    users: createSeedUsers(),
    tracks: createSeedTracks(),
    albums: createSeedAlbums(),
    playlists: createSeedPlaylists(),
    notifications: createSeedNotifications(),
    tickets: createSeedTickets(),
    payouts: createSeedPayouts(),
    prices: createSeedPrices(),
  };
}

File: backend/accounts/management/commands/seed.py
from datetime import date, timedelta

from django.core.management.base import BaseCommand
from django.utils import timezone

from accounts.models import ArtistProfile, Follow, User
from billing.models import ArtistPayout, Payment, SubscriptionPrice
from catalog.models import Album, Playlist, PlaylistTrack, StreamEvent, Track
from social.models import Notification, Ticket, TicketMessage

AUDIO = {
    "a": "https://www.soundhelix.com/examples/mp3/SoundHelix-Song-1.mp3",
    "b": "https://www.soundhelix.com/examples/mp3/SoundHelix-Song-2.mp3",
    "c": "https://www.soundhelix.com/examples/mp3/SoundHelix-Song-3.mp3",
    "d": "https://www.soundhelix.com/examples/mp3/SoundHelix-Song-4.mp3",
    "e": "https://www.soundhelix.com/examples/mp3/SoundHelix-Song-5.mp3",
}


def cover(seed: str) -> str:
    return f"https://picsum.photos/seed/{seed}/400/400"


class Command(BaseCommand):
    help = "Load demo users, catalog, tickets, and prices."

    def handle(self, *args, **options):
        if User.objects.filter(email="admin@demo.com").exists():
            self.stdout.write("Seed already present — skipped.")
            return

        day = date.today()
        password = "demo1234"

        sara = User.objects.create_user(
            email="sara@demo.com",
            password=password,
            username="sara_basic",
            display_name="سارا شنونده",
            role=User.Role.LISTENER,
            gender=User.Gender.FEMALE,
            birth_date=date(1998, 5, 12),
            subscription_tier=User.Tier.BASIC,
            daily_stream_count=12,
            daily_stream_date=day,
        )
        ali = User.objects.create_user(
            email="ali@demo.com",
            password=password,
            username="ali_silver",
            display_name="علی نقره‌ای",
            role=User.Role.LISTENER,
            gender=User.Gender.MALE,
            birth_date=date(1995, 3, 20),
            subscription_tier=User.Tier.SILVER,
            subscription_expires_at=date(2026, 12, 1),
            daily_stream_count=40,
            daily_stream_date=day,
        )
        nima = User.objects.create_user(
            email="nima@demo.com",
            password=password,
            username="nima_gold",
            display_name="نیما طلایی",
            role=User.Role.LISTENER,
            gender=User.Gender.MALE,
            birth_date=date(1992, 11, 8),
            subscription_tier=User.Tier.GOLD,
            subscription_expires_at=date(2027, 1, 1),
            daily_stream_count=100,
            daily_stream_date=day,
        )
        ava = User.objects.create_user(
            email="ava@demo.com",
            password=password,
            username="ava_mehr",
            display_name="آوا مهر",
            role=User.Role.ARTIST,
            subscription_tier=User.Tier.GOLD,
            daily_stream_date=day,
        )
        ArtistProfile.objects.create(
            user=ava,
            bio="خواننده و آهنگساز مستقل با تمرکز روی پاپ و الکترونیک.",
            verified=True,
            status=ArtistProfile.Status.APPROVED,
            sample_works="نمونه‌کارهای منتشرشده روی پلتفرم",
            stage_name="آوا مهر",
        )
        kaveh = User.objects.create_user(
            email="kaveh@demo.com",
            password=password,
            username="kaveh_new",
            display_name="کاوه نوپا",
            role=User.Role.ARTIST,
            daily_stream_date=day,
        )
        ArtistProfile.objects.create(
            user=kaveh,
            bio="هنرمند تازه‌کار در انتظار تأیید.",
            verified=False,
            status=ArtistProfile.Status.PENDING,
            sample_works="لینک دمو / فایل‌های نمونه",
            stage_name="کاوه نوپا",
        )
        User.objects.create_user(
            email="support@demo.com",
            password=password,
            username="support1",
            display_name="پشتیبان ملودی",
            role=User.Role.SUPPORT,
            subscription_tier=User.Tier.GOLD,
            daily_stream_date=day,
        )
        User.objects.create_user(
            email="admin@demo.com",
            password=password,
            username="admin",
            display_name="مدیر سامانه",
            role=User.Role.ADMIN,
            is_staff=True,
            is_superuser=True,
            subscription_tier=User.Tier.GOLD,
            daily_stream_date=day,
        )

        Follow.objects.bulk_create(
            [
                Follow(follower=sara, following=ava),
                Follow(follower=ali, following=ava),
                Follow(follower=nima, following=ava),
                Follow(follower=nima, following=kaveh),
                Follow(follower=nima, following=sara),
                Follow(follower=sara, following=nima),
            ]
        )

        album1 = Album.objects.create(
            title="شهر خاموش",
            year=2025,
            genre="پاپ",
            cover_url=cover("album1"),
        )
        album1.artists.add(ava)
        album2 = Album.objects.create(
            title="ارتفاع",
            year=2024,
            genre="راک",
            cover_url=cover("album2"),
        )
        album2.artists.add(ava)

        tracks_data = [
            ("سپیده‌دم", album1, AUDIO["a"], cover("track1"), "پاپ", "در سپیده‌دم\nصدای شهر آرام می‌گیرد", 372, 15200, 4200, False, False),
            ("شب‌های تهران", album1, AUDIO["b"], cover("track2"), "پاپ", "شب‌های تهران\nنور نئون و صدای باران", 401, 22100, 6100, False, False),
            ("موج آرام", album1, AUDIO["c"], cover("track3"), "الکترونیک", "", 350, 9800, 3000, False, False),
            ("تک‌آهنگ طلایی", None, AUDIO["d"], cover("track4"), "ایندی", "این تک‌آهنگ زودهنگام است", 310, 1200, 800, True, True),
            ("پرواز", album2, AUDIO["e"], cover("track5"), "راک", "", 290, 45000, 12000, False, False),
        ]
        created_tracks = []
        for title, album, audio, cov, genre, lyrics, duration, streams, listeners, early, single in tracks_data:
            track = Track.objects.create(
                title=title,
                album=album,
                audio_url=audio,
                cover_url=cov,
                genre=genre,
                lyrics=lyrics,
                year=2025 if album != album2 else 2024,
                duration=duration,
                streams=streams,
                listeners=listeners,
                early_access=early,
                is_single=single or album is None,
            )
            track.artists.add(ava)
            created_tracks.append(track)

        pl1 = Playlist.objects.create(owner=sara, name="تمرکز صبح", cover_url=cover("pl1"))
        PlaylistTrack.objects.create(playlist=pl1, track=created_tracks[0], position=0)
        PlaylistTrack.objects.create(playlist=pl1, track=created_tracks[2], position=1)
        pl2 = Playlist.objects.create(owner=nima, name="شب‌گردی", cover_url=cover("pl2"))
        for i, track in enumerate([created_tracks[1], created_tracks[3], created_tracks[4]]):
            PlaylistTrack.objects.create(playlist=pl2, track=track, position=i)

        now = timezone.now()
        for i in range(8):
            StreamEvent.objects.create(
                user=nima, track=created_tracks[i % 5], created_at=now - timedelta(days=i)
            )
            StreamEvent.objects.create(
                user=ali, track=created_tracks[0], created_at=now - timedelta(hours=i + 1)
            )

        Notification.objects.bulk_create(
            [
                Notification(
                    user=sara,
                    title="مهلت اشتراک",
                    body="اشتراک پایه شما رایگان است؛ برای امکانات بیشتر ارتقا دهید.",
                    href="/settings",
                ),
                Notification(
                    user=nima,
                    title="اثر جدید",
                    body="آوا مهر تک‌آهنگ جدیدی منتشر کرد.",
                    href="/browse",
                ),
                Notification(
                    user=kaveh,
                    title="در انتظار تأیید",
                    body="درخواست حساب هنرمندی شما در حال بررسی است.",
                    read=True,
                ),
                Notification(
                    user=User.objects.get(email="support@demo.com"),
                    title="تیکت جدید",
                    body="یک تیکت پشتیبانی جدید ثبت شد.",
                    href="/admin/tickets",
                ),
                Notification(
                    user=User.objects.get(email="admin@demo.com"),
                    title="درخواست هنرمند",
                    body="کاوه نوپا درخواست تأیید حساب داده است.",
                    href="/admin/artist-requests",
                ),
            ]
        )

        t1 = Ticket.objects.create(user=sara, subject="مشکل پخش آهنگ", status=Ticket.Status.OPEN)
        TicketMessage.objects.create(
            ticket=t1, sender=sara, body="بعضی آهنگ‌ها بعد از چند ثانیه قطع می‌شوند."
        )
        t2 = Ticket.objects.create(user=ali, subject="سوال درباره دانلود", status=Ticket.Status.ANSWERED)
        TicketMessage.objects.create(
            ticket=t2, sender=ali, body="دانلود آهنگ از کجا فعال می‌شود؟"
        )
        TicketMessage.objects.create(
            ticket=t2,
            sender=User.objects.get(email="support@demo.com"),
            body="از منوی کارت آهنگ گزینه دانلود را بزنید (اشتراک نقره‌ای و بالاتر).",
        )

        prices = SubscriptionPrice.get_solo()
        prices.silver = 99000
        prices.gold = 199000
        prices.save()

        Payment.objects.create(
            user=nima,
            tier=Payment.Tier.GOLD,
            months=12,
            amount=199000 * 12,
            status=Payment.Status.SUCCESS,
            authority="SEED-GOLD",
            gateway="mock",
            ref_id="SEED-REF",
        )

        ArtistPayout.objects.create(
            artist=ava,
            unique_listeners=8200,
            streams=94000,
            amount=8200 * 500 + 94000 * 50,
            payment_status=ArtistPayout.PaymentStatus.PENDING,
            month="2026-07",
        )

        self.stdout.write(self.style.SUCCESS("Demo data seeded."))
